File: Prototype/BigGp/src/Model/BaseNodes/Assignment/Assignment.java
package Model.BaseNodes.Assignment;

import Model.BaseNodes.AddRandomBaseNode;
import Model.Interfaces.SubtreeMutable;
import Model.Node;
import Model.SmallNodes.Expressions.Standart.Childs.VarExpression;
import Model.SmallNodes.Expressions.Standart.Expression;
import Serialization.Token;

import java.util.ArrayList;
import java.util.Random;


public class Assignment extends Node implements SubtreeMutable {

    //    Currently hardcoded, that we can only get expression not read() during program generation
//    If we want to use read() we have to use one of 2 constructors
    enum AssignmentChild {
        EXPRESSION,
        FUNCTION_CALL_IN;
        private static final AssignmentChild[] VALUES = AssignmentChild.class.getEnumConstants();
        private static final int SIZE = VALUES.length;
        private static final Random RANDOM = new Random();

        public static AssignmentChild randomChild() {
            return VALUES[(RANDOM.nextInt(SIZE))];
        }
    }

    private VarExpression variable;
    private Expression expression;
    private FunctionCallIn read;

    @Override
    public String toString() {
        if (read == null) {
            return variable.getVariableName() + " = " + expression.toString() + ";";
        } else {
            return variable.getVariableName() + " = " + read.toString() + ";";
        }
    }

    public void setExpression(Expression expression) {
        this.expression = expression;
    }

    @Override
    public void subtreeMutate() {
        Node parentNode = this.getParentNode();
        double nextDeepNodeChance = ((AddRandomBaseNode) parentNode).getNextDeepNodeChance();
        Node node = ((AddRandomBaseNode) parentNode).getRandomNode(parentNode, nextDeepNodeChance);

        ((AddRandomBaseNode) parentNode).reAttachSubtree(this, node);
    }

    public FunctionCallIn getRead() {
        return read;
    }

    public VarExpression getVariable() {
        return variable;
    }

    public ArrayList<Node> getChildrenAsNodes() {
        ArrayList<Node> nodes = new ArrayList<>();
        nodes.add(this);
        nodes.addAll(variable.getChildrenAsNodes());

        if (read == null) {
            nodes.addAll(expression.getChildrenAsNodes());
        } else {
            nodes.addAll(read.getChildrenAsNodes());
        }

        return nodes;
    }

    public Assignment(Node parent) {
        super(parent, "Assignment", true);

//        Hardcoded for expression
        expression = new Expression(this);
        read = null;

        variable = new VarExpression(this);

//        AssignmentChild assignmentChild = AssignmentChild.randomChild();
//
//        switch (assignmentChild) {
//            case EXPRESSION -> {
//                value = new Expression(this);
//                read = null;
//            }
//            case FUNCTION_CALL_IN -> {
//                read = new FunctionCallIn(this);
//                value = null;
//            }
//        }
    }

    //    USE THIS ONLY DURING PROGRAM INPUT DECLARATION!!!!!!
    public Assignment(Node parent, boolean shouldVarAlreadyExist) {
        super(parent, "Assignment", true);

        try {
            variable = new VarExpression(this, shouldVarAlreadyExist);
        } catch (Exception e) {
            e.printStackTrace();
            System.out.println(
                    "[shouldVarAlreadyExist] value was" + shouldVarAlreadyExist +
                            "when u didnt have any variables existing." +
                            "This should never happen bcs this constructor" +
                            "was ment to be used only during program input declarations when" +
                            "[shouldVarAlreadyExist] value should always be false!!");
        }

        read = new FunctionCallIn(this);
        expression = null;
    }

    private void assignmentFromTokens(ArrayList<Token> tokens) {
        switch (tokens.get(0).getNAME()) {
            case "VarExpression" -> variable = new VarExpression(this, tokens);
            case "Expression" -> expression = new Expression(this, tokens);
            case "FunctionCallIn" -> read = new FunctionCallIn(this, tokens);
        }
    }

    public Assignment(Node parent, ArrayList<Token> tokens) {
        super(parent, "Assignment", true);
        tokens.remove(0);

        assignmentFromTokens(tokens);
        assignmentFromTokens(tokens);
    }
}

File: Interactive.Genetic.Programming.Environment/src/Domain/Model/Nodes.Big/Assignments/Assignment.cs
using CommunityToolkit.Diagnostics;
using Model.Abstract;
using Model.Interfaces;
using Model.Nodes.Small.Expressions.Standard;

namespace Model.Nodes.Big.Assignments;

public enum AssignmentChild
{
    Expression,
    FunctionCallIn
}

public sealed class Assignment : Node, ISubtreeMutable
{
    public VarExpression Variable { get; set; }
    public StandardExpression? Expression { get; set; }
    public FunctionCallIn? Read { get; set; }

    public override List<Node> ChildrenAsNodes()
    {
        var nodes = new List<Node> { this };
        nodes.AddRange(Variable.ChildrenAsNodes());
        nodes.AddRange(
            Expression?.ChildrenAsNodes()
            ?? Read?.ChildrenAsNodes()
            ?? throw new InvalidOperationException("Both expression and read are null")
        );
        return nodes;
    }

    public void SubtreeMutate()
    {
        var parent = ParentNode as BigNode ?? throw new InvalidOperationException("Parent is not a big node");
        var node = parent.GetRandomNode();
        parent.ReAttachSubtree(this, node);
    }

    private void FromTokens(List<Token> tokens)
    {
        switch (tokens[0].Name)
        {
            case "VarExpression":
                Variable = new VarExpression(this, tokens);
                break;
            case "Expression":
                Expression = new StandardExpression(this, tokens);
                break;
            case "FunctionCallIn":
                Read = new FunctionCallIn(this, tokens);
                break;
        }
    }

    // hardcoded for expression
    public Assignment(Node parentNode) : base(parentNode, "Assignment", true)
    {
        Expression = new StandardExpression(this);
        Variable = new VarExpression(this, true);
    }

    // only during program input declaration
    public Assignment(Node parentNode, bool shouldVarAlreadyExist) : base(parentNode, "Assignment", true)
    {
        Guard.IsFalse(shouldVarAlreadyExist);
        Variable = new VarExpression(this, shouldVarAlreadyExist);
        Read = new FunctionCallIn(this);
    }

    public Assignment(Node parentNode, List<Token> tokens) : base(parentNode, "Assignment", true)
    {
        tokens.RemoveAt(0);
        FromTokens(tokens);
        FromTokens(tokens);
    }

    public override string ToString()
        => Read is null
            ? $"{Variable.Name} = {Expression}"
            : $"{Variable.Name} = {Read}";
}